File: bad_code_from_long_ago/cf101510/c.cpp
#include <bits/stdc++.h>
using namespace std;
#define _USE_MATH_DEFINES

typedef long long ll;
typedef long double ld;
typedef pair<int,int> pii;
typedef complex<ld> pt;
typedef vector<pt> pol;
#define nl '\n'

inline void srand() { srand(clock() + time(nullptr)); }

template <class T, class U>
ostream& operator << (ostream& os, const pair<T,U>& v) {
	os << "(" << v.first << "," << v.second << ")"; return os;
}

template <class T>
ostream& operator << (ostream& os, const vector<T>& v) {
	for (const T& it : v) os << it << " "; return os;
}

template <class T>
inline T sqr(T x) { return x*x; }

const ll INF = 0x3f3f3f3f;
const ll INFLL = 0x3f3f3f3f3f3f3f3f;
//const ld PI = M_PIl;
////////////////////////////////////////////////////////////////////////

const ll MOD = 1e9+7;
const ld EPS = 1e-9;

bool valid(int n, int a[], int len, int k) {
	int good = -1;
	for (int i = 0; i < n; i++) {
		int cur = upper_bound(a, a+n, a[i]+len) - (a+i);
		if (cur < k && good < i) return false;
		else if (cur >= k) good = i+cur-1;
	}
	return true;
}

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

	int n, k;
	cin >> n >> k;

	int a[n];
	for (int i = 0; i < n; i++) {
		cin >> a[i];
	}
	sort(a, a+n);

	ll left, right, ans;
	left = 0; right = 2e9+1;
	while (left <= right) {
		ans = (left + right) / 2;
		if (valid(n, a, ans, k))
			right = ans-1;
		else
			left = ans+1;
	}
	if (!valid(n, a, ans, k)) {
		ans++;
	}
	cout << ans << nl;

	return 0;
}

File: bad_code_from_long_ago/cf101510/c.rb
def valid(arr, len, k)
  good = -1
  n = arr.length
  arr.push(0x3f3f3f3f3f3f3f3f3f)
  idarr = 0..n
  for i in 0..n-1
    # cur = arr.bsearch_index{|x| x > arr[i]+len} - i
    cur = idarr.bsearch{|x| arr[x] > arr[i]+len} - i
    if cur < k and good < i
      return false
    elif cur >= k
      good = i+cur-1
    end
  end
  return true
end

n, k = readline().split().map(&:to_i)
arr = readline().split().map(&:to_i).sort()

left = 0
right = 2*10**9+1
ans = 0
while left <= right
  ans = (left+right)/2
  if valid(arr, ans, k)
    right = ans-1
  else
    left = ans+1
  end
end

if not valid(arr, ans, k)
  ans += 1
end

puts ans
